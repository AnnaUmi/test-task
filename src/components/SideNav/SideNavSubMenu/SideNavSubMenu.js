import React, { Component } from 'react';
import PropTypes from 'prop-types';

import SideNavSubMenuListItem from '../SideNavSubMenuListItem';
import { SideNavListItemStyled, SubMenuListItemContainer } from '../SideNav.styles';
import { sideNavSubMenuItemProps } from '../../../constants';

class SideNavSubMenu extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isOpened: this.isOpenedOnInit(),
    };
  }

  isOpenedOnInit = () => {
    const { listItems, currRoute } = this.props;

    return listItems.reduce((acc, curr) => {
      if (curr.url === currRoute) {
        return true;
      }

      return acc;
    }, false);
  };

  toggleMenu = () => {
    const { isOpened } = this.state;
    this.setState({
      isOpened: !isOpened,
    });
  };

  render() {
    const { listItems, label, currRoute } = this.props;
    const { isOpened } = this.state;
    return (
      <div>
        <SideNavListItemStyled onClick={this.toggleMenu} isActive={isOpened}>
          {label}
        </SideNavListItemStyled>
        <SubMenuListItemContainer isOpened={isOpened}>
          {listItems.map(item => (
            <SideNavSubMenuListItem isActive={currRoute === item.url} {...item} key={item.id} />
          ))}
        </SubMenuListItemContainer>
      </div>
    );
  }
}

SideNavSubMenu.propTypes = {
  label: PropTypes.string.isRequired,
  listItems: PropTypes.arrayOf(PropTypes.shape(sideNavSubMenuItemProps)).isRequired,
  currRoute: PropTypes.string.isRequired,
};

export default SideNavSubMenu;
