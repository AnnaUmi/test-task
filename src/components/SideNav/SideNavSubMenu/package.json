{
  "name": "side-nav-sub-menu",
  "version": "0.0.0",
  "private": true,
  "main": "./SideNavSubMenu.js"
}
