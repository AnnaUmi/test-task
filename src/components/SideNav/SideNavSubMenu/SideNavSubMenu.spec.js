import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import SideNavSubMenu from './SideNavSubMenu';

const listItems = [
  {
    label: 'test sub item',
    url: 'test/sub/item',
    id: 'test/sub/item',
  },
];

describe('SideNavSubMenuListItem', () => {
  it('should match snapshot when not active', () => {
    const tree = renderer.create(
      <SideNavSubMenu label="test label" listItems={listItems} currRoute="no/match" />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should match snapshot when is active', () => {
    const tree = renderer.create(
      <SideNavSubMenu label="test label" listItems={listItems} currRoute="test/sub/item" />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  it('should call setState when toggleMenu is fired', () => {
    const tree = renderer.create(
      <SideNavSubMenu label="test label" listItems={listItems} currRoute="test/sub/item" />
    );
    const instance = tree.getInstance();
    const setStateMock = jest.fn();

    instance.setState = setStateMock;
    instance.toggleMenu();

    expect(setStateMock).toHaveBeenCalled();
    expect(setStateMock).toHaveBeenCalledTimes(1);
  });
});
