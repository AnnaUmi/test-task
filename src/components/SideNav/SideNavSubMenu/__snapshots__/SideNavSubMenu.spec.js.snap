// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideNavSubMenuListItem should match snapshot when is active 1`] = `
.c0 {
  padding: 20px 30px;
  font-size: 20px;
  cursor: pointer;
  background-color: #7e7e7e;
  font-weight: bold;
}

.c0:hover {
  background-color: #bdbdbd;
}

.c2 {
  padding: 20px 60px;
  font-size: 18px;
  cursor: pointer;
  background-color: #7e7e7e;
  font-weight: bold;
}

.c2:hover {
  background-color: #bdbdbd;
}

.c1 {
  background-color: #d3d3d3;
}

<div>
  <div
    className="c0"
    onClick={[Function]}
  >
    test label
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
      onClick={[Function]}
    >
      test sub item
    </div>
  </div>
</div>
`;

exports[`SideNavSubMenuListItem should match snapshot when not active 1`] = `
.c0 {
  padding: 20px 30px;
  font-size: 20px;
  cursor: pointer;
}

.c0:hover {
  background-color: #bdbdbd;
}

.c2 {
  padding: 20px 60px;
  font-size: 18px;
  cursor: pointer;
}

.c2:hover {
  background-color: #bdbdbd;
}

.c1 {
  background-color: #d3d3d3;
  height: 0;
  overflow: hidden;
}

<div>
  <div
    className="c0"
    onClick={[Function]}
  >
    test label
  </div>
  <div
    className="c1"
  >
    <div
      className="c2"
      onClick={[Function]}
    >
      test sub item
    </div>
  </div>
</div>
`;
